File: src/main/java/net/yakclient/mixin/registry/pool/ExternalLibRegistryPool.java
package net.yakclient.mixin.registry.pool;

import net.yakclient.mixin.internal.loader.PackageTarget;

import java.net.URL;

public class ExternalLibRegistryPool extends RegistryPool<URL> {
    @Override
    public Location pool(URL type) {
        final ExternalLibLocation key = new ExternalLibLocation(type);
        if (!this.pool.containsKey(key)) this.pool.put(key, new PoolQueue<>());

        this.pool.get(key).add(type, (t)->{}, null);

        return key;
    }

    @Override
    public PackageTarget register(Location location) {
        return null;
    }
}

File: base/src/main/kotlin/net/yakclient/mixin/base/registry/pool/ExternalLibRegistryPool.kt
package net.yakclient.mixin.base.registry.pool

import net.yakclient.mixin.base.internal.loader.ContextPoolManager
import net.yakclient.mixin.base.registry.proxy.MixinProxyManager
import net.yakclient.mixin.base.target.JarTarget
import net.yakclient.mixin.base.target.Target
import java.net.URL

class ExternalLibRegistryPool : RegistryPool<URL>() {
    override fun pool(type: URL): Location {
        return ExternalLibLocation(type).also { loc ->
            (if (!pool.containsKey(loc)) PoolQueue<URL>().also { pool[loc] = it } else pool[loc]!!).add(type)
            if (!pool.containsKey(loc)) pool[loc] = PoolQueue<URL>().apply { add(type) }
        }
    }

    override fun mix(location: Location) {
        require(location is ExternalLibLocation) { "Given location must be a external library to register with the lib registry pool!" }
        ContextPoolManager.libPool.addTarget(JarTarget(location.url))
    }
}